Rubberduck.Inspections/Concrete/ShadowedDeclarationInspection.cs

```diff
@@ -0,0 +1,628 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Linq;
+using Antlr4.Runtime;
+using Antlr4.Runtime.Tree;
+using Rubberduck.Common;
+using Rubberduck.Inspections.Abstract;
+using Rubberduck.Inspections.Results;
+using Rubberduck.Parsing;
+using Rubberduck.Parsing.Grammar;
+using Rubberduck.Parsing.Inspections.Abstract;
+using Rubberduck.Parsing.Inspections.Resources;
+using Rubberduck.Parsing.Symbols;
+using Rubberduck.Parsing.VBA;
+using Rubberduck.UI;
+using Rubberduck.VBEditor.SafeComWrappers;
+
+namespace Rubberduck.Inspections.Concrete
+{
+    public sealed class ShadowedDeclarationInspection : InspectionBase
+    {
+        private enum DeclarationSite
+        {
+            NotApplicable = 0,
+            ReferencedProject = 1,
+            OtherComponent = 2,
+            SameComponent = 3
+        }
+
+        private class OptionPrivateModuleListener : VBAParserBaseListener
+        {
+            public List<VBAParser.ModuleContext> OptionPrivateModules { get; } = new List<VBAParser.ModuleContext>();
```

Comment:
I recently learned that we should be using visitors, not listeners, because visitors can return results whereas listeners don't. I would refactor the listeners into visitors so that we have a common approach for extracting contexts/tokens.

Reply:
I'll take the listener initial version, and we can refactor to visitors in a further iteration? I've been wanting to merge this PR for weeks!